drv/psc-psu-update/src/main.rs

```diff
@@ -0,0 +1,389 @@
+                    break;
+                }
+
+                Ok((UpdateState::UpdateSuccessful, _)) => {
+                    let state = Some(UpdateState::UpdateSuccessful);
+
+                    //
+                    // We should be back up!  As a final measure, we are going
+                    // to check that the firmware revision matches the
+                    // revision we think we just wrote.  If it doesn't, there
+                    // is something amiss:  it may be that the image is
+                    // corrupt or that the version doesn't otherwise match.
+                    // Regardless, we consider that to be an update failure.
+                    //
+                    match dev.firmware_revision() {
+                        Ok(revision) if revision != MWOCP68_FIRMWARE_REV => {
+                            update_failed(state, None);
+                            break;
+                        }
+
+                        Err(err) => {
+                            update_failed(state, Some(err));
+                            break;
+                        }
+
+                        Ok(_) => {}
+                    }
+
+                    //
+                    // We're on the new firmware!  And now, a final final
+                    // check: make sure that we are power-good.  It is very
+                    // unclear what to do here if are NOT power-good:  we know
+                    // that we WERE power-good before we started, so it
+                    // certainly seems possible that we have put a firmware
+                    // update on this PSU which has somehow incapacitated it.
+                    // We would rather not put the system in a compromised
+                    // state by continuing to potentially brick PSUs -- but we
+                    // also want to assure that we make progress should this
+                    // ever resolve (e.g., by pulling the bricked PSU). We will
+                    // remain here until we see the updated PSU go power-good;
+                    // if it never does, we will at least not attempt to put
+                    // the (potentially) bad update anywhere else!
+                    //
+                    loop {
+                        match dev.power_good() {
+                            Ok(power_good) if power_good => break,
+                            Ok(_) => {
+                                ringbuf_entry!(Trace::PowerNotGood(ndx));
+                            }
+                            Err(err) => {
+                                ringbuf_entry!(Trace::PGError(ndx, err));
+                            }
+                        }
+
+                        hl::sleep_for(TIMER_INTERVAL_MS);
+                    }
+
+                    ringbuf_entry!(Trace::UpdateSucceeded);
+                    self.update_succeeded = Some(Ticks::now());
+                    self.update_backoff = None;
```

Comment:
Clear `serial_number` here as well, to avoid confusion when moving to the next PSU?

Reply:
Whoops, never mind, this is tracked on a per-PSU basis